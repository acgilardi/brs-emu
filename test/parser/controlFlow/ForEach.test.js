const Parser = require("../../../lib/parser");
const Expr = require("../../../lib/parser/Expression");
const Stmt = require("../../../lib/parser/Statement");
const { Lexeme } = require("../../../lib/Lexeme");
const BrsError = require("../../../lib/Error");
const { BrsString } = require("../../../lib/brsTypes");

const { EOF } = require("../ParserTests");

describe("parser foreach loops", () => {
    afterEach(() => BrsError.reset());

    it("requires a name and target", () => {
        let parsed = Parser.parse([
            { kind: Lexeme.ForEach, text: "for each", line: 2 },
            { kind: Lexeme.Identifier, text: "word", line: 2 },
            { kind: Lexeme.Identifier, text: "in", line: 2 },
            { kind: Lexeme.Identifier, text: "lipsum", line: 2 },
            { kind: Lexeme.Newline, text: "\n", line: 2 },

            // body would go here, but it's not necessary for this test
            { kind: Lexeme.EndFor, text: "end for", line: 3 },
            { kind: Lexeme.Newline, text: "\n", line: 3 },
            EOF
        ]);

        expect(BrsError.found()).toBe(false);
        expect(parsed).toBeDefined();

        let forEach = parsed[0];
        expect(forEach).toBeInstanceOf(Stmt.ForEach);

        expect(forEach.item).toEqual(
            { kind: Lexeme.Identifier, text: "word", line: 2 }
        );
        expect(forEach.target).toBeInstanceOf(Expr.Variable);
        expect(forEach.target.name).toEqual(
            { kind: Lexeme.Identifier, text: "lipsum", line: 2 }
        );

        expect(parsed).toMatchSnapshot();
    });
});