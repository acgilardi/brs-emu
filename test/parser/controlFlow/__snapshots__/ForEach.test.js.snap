// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser foreach loops requires a name and target 1`] = `
Array [
  ForEach {
    "body": Block {
      "statements": Array [],
    },
    "item": Object {
      "kind": 21,
      "line": 2,
      "text": "word",
    },
    "target": Variable {
      "name": Object {
        "kind": 21,
        "line": 2,
        "text": "lipsum",
      },
    },
  },
]
`;
